internal/rules/resources.go

```diff
@@ -0,0 +1,123 @@
+package rules
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/titpetric/factory"
+
+	"github.com/crusttech/crust/internal/auth"
+	"github.com/crusttech/crust/system/types"
+)
+
+type Access string
+
+const (
+	Allow   Access = "yes"
+	Deny           = "no"
+	Inherit        = ""
+)
+
+type resources struct {
+	ctx context.Context
+	db  *factory.DB
+}
+
+func NewResources(ctx context.Context, db *factory.DB) ResourcesInterface {
+	return (&resources{}).With(ctx, db)
+}
+
+func (r *resources) With(ctx context.Context, db *factory.DB) ResourcesInterface {
+	return &resources{
+		ctx: ctx,
+		db:  db,
+	}
+}
+
+func (r *resources) identity() uint64 {
+	return auth.GetIdentityFromContext(r.ctx).Identity()
+}
+
+func (r *resources) CheckAccessMulti(resource string, operation string) error {
+	user := r.identity()
+	result := []Access{}
+	query := []string{
+		// select rules
+		"select r.value from sys_rules r",
+		// join members
+		"inner join sys_team_member m on (m.rel_team = r.rel_team and m.rel_user=?)",
+		// add conditions
+		"where r.resource LIKE ? and r.operation=?",
+	}
+	resource = strings.Replace(resource, "*", "%", -1)
+	queryString := strings.Join(query, " ")
+	if err := r.db.Select(&result, queryString, user, resource, operation); err != nil {
+		return err
+	}
+
+	// order by deny, allow
+	for _, val := range result {
+		if val == Deny {
+			return errors.New("Access not allowed")
+		}
+	}
+	for _, val := range result {
+		if val == Allow {
+			return nil
+		}
+	}
+	return errors.New("Access not allowed")
+}
+
+func (r *resources) CheckAccess(resource string, operation string) error {
+	user := r.identity()
+	result := []Access{}
+	query := []string{
+		// select rules
+		"select r.value from sys_rules r",
+		// join members
+		"inner join sys_team_member m on (m.rel_team = r.rel_team and m.rel_user=?)",
+		// add conditions
+		"where r.resource=? and r.operation=?",
+	}
+	queryString := strings.Join(query, " ")
+	if err := r.db.Select(&result, queryString, user, resource, operation); err != nil {
+		return err
+	}
+
+	// order by deny, allow
+	for _, val := range result {
+		if val == Deny {
+			return errors.New("Access not allowed")
```

Comment:
is this really an error? shouldn't this return bool only?

Reply:
Yes and no. Errors are easier to handle, but a TODO which you can see here is to add error types so we can check higher up if the returned error is a ErrDeniedAccess or something else (database issue, or with didmos a http response issue)...

tl;dr: introduce error type for ErrDeniedAccess; and maybe (@mitjaziv opinion) a helper function that just logs the error and returns a bool?